arnglld/src/main.rs

```diff
@@ -33,19 +40,175 @@ struct Opt {
     #[clap(short, long, parse(from_occurrences))]
     verbose: usize,
 
+    /// List audio devices
+    #[clap(long)]
+    list_devices: bool,
+
     #[clap(short, long)]
-    callsign: HamAddr,
+    callsign: Option<HamAddr>,
+
+    #[clap(long)]
+    input_audio_device: Option<String>,
+
+    #[clap(long)]
+    output_audio_device: Option<String>,
+}
+
+impl Opt {
+    fn get_output_device(&self) -> Result<cpal::Device, anyhow::Error> {
+        let host = cpal::default_host();
+
+        if let Some(name) = self.output_audio_device.as_ref() {
+            let lc_name = name.to_lowercase();
+
+            // First try exact matches
+            for device in host.output_devices()? {
```

Comment:
There's a lot of repetition in this method.  Perhaps some of it could be collapsed?

let output_devices = host.output_devices()?;
let devs = output_devices.iter().map(|d| (d.name()?.to_lowercase(), d)).collect();
// Try exact
if let Some((_, device)) = devs.find(|(n, _)| n == lc_name) { return Ok(device); }
// Try prefix
if let Some((_, device)) = devs.find(|n, _)| n.starts_with(&lc_name)) { return Ok(device); }
// Any substring match
if let Some((_, device)) = devs.find(|n, _)| n.contains(&lc_name)) { return Ok(device); }
Err(...)

Reply:
Yep, I like that better. Will fix.

Reply:
I rewrote to use `filter_map` to avoid needing the `flatten` step.